2000-10-10  Jakub Jelinek  <jakub@redhat.com>

	* expmed.c (store_bit_field): Set SUBREG_BYTE correctly on big
	endian.
	(extract_bit_field): Likewise.

--- gcc/expmed.c.jj	Fri Sep 15 16:46:46 2000
+++ gcc/expmed.c	Tue Oct 10 18:05:42 2000
@@ -327,7 +327,8 @@ store_bit_field (str_rtx, bitsize, bitnu
 	    }
 	  if (GET_CODE (op0) == REG)
 	    op0 = gen_rtx_SUBREG (fieldmode, op0,
-				  (offset * UNITS_PER_WORD));
+				  (bitnum % BITS_PER_WORD) / BITS_PER_UNIT
+				  + (offset * UNITS_PER_WORD));
 	  else
 	    op0 = change_address (op0, fieldmode,
 				  plus_constant (XEXP (op0, 0), offset));
@@ -372,7 +373,9 @@ store_bit_field (str_rtx, bitsize, bitnu
 	}
 
       emit_insn (GEN_FCN (icode)
-		 (gen_rtx_SUBREG (fieldmode, op0, (offset * UNITS_PER_WORD)),
+		 (gen_rtx_SUBREG (fieldmode, op0,
+				  (bitnum % BITS_PER_WORD) / BITS_PER_UNIT
+				  + (offset * UNITS_PER_WORD)),
 				  value));
 
       return value;
@@ -1095,7 +1098,9 @@ extract_bit_field (str_rtx, bitsize, bit
 		abort ();
 	    }
 	  if (GET_CODE (op0) == REG)
-	    op0 = gen_rtx_SUBREG (mode1, op0, (offset * UNITS_PER_WORD));
+	    op0 = gen_rtx_SUBREG (mode1, op0,
+				  (bitnum % BITS_PER_WORD) / BITS_PER_UNIT
+				  + (offset * UNITS_PER_WORD));
 	  else
 	    op0 = change_address (op0, mode1,
 				  plus_constant (XEXP (op0, 0), offset));
